cmd/controller/main.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/metrics"
+	pkgmetrics "github.com/knative/pkg/metrics"
+	"github.com/knative/pkg/signals"
+	"go.uber.org/zap"
+)
+
+const (
+	component = "controller"
+
+	// raImageEnvVar is the name of the environment variable that contains the receive adapter's
+	// image. It must be defined.
+	raPubSubImageEnvVar = "GCPPUBSUB_RA_IMAGE"
+)
+
+var (
+	masterURL  = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+)
+
+func main() {
+	flag.Parse()
+
+	// Set up our logger.
+	loggingConfigMap, err := configmap.Load("/etc/config-logging")
+	if err != nil {
+		log.Fatal("Error loading logging configuration:", err)
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatal("Error parsing logging configuration:", err)
+	}
+	logger, atomicLevel := logging.NewLoggerFromConfig(loggingConfig, component)
+	defer flush(logger)
+
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		logger.Fatalw("Error building kubeconfig", zap.Error(err))
+	}
+
+	const numControllers = 1
+	cfg.QPS = numControllers * rest.DefaultQPS
+	cfg.Burst = numControllers * rest.DefaultBurst
+	opt := reconciler.NewOptionsOrDie(cfg, logger, stopCh)
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(opt.KubeClientSet, opt.ResyncPeriod)
+	runInformerFactory := informers.NewSharedInformerFactory(opt.RunClientSet, opt.ResyncPeriod)
+
+	deploymentInformer := kubeInformerFactory.Apps().V1().Deployments()
+	pubsubSourceInformer := runInformerFactory.Events().V1alpha1().PubSubSources()
+
+	raPubSubSourceImage, defined := os.LookupEnv(raPubSubImageEnvVar)
+	if !defined {
+		logger.Fatalw("required environment variable '%s' not defined", raPubSubImageEnvVar)
+	}
+
+	// Build all of our controllers, with the clients constructed above.
+	// Add new controllers to this array.
+	controllers := [...]*controller.Impl{
+		pubsubsource.NewController(
+			opt,
+			deploymentInformer,
+			pubsubSourceInformer,
+			raPubSubSourceImage,
+		),
+	}
+	// This line asserts at compile time that the length of controllers is equal to numControllers.
+	// It is based on https://go101.org/article/tips.html#assert-at-compile-time, which notes that
+	// var _ [N-M]int
+	// asserts at compile time that N >= M, which we can use to establish equality of N and M:
+	// (N >= M) && (M >= N) => (N == M)
+	var _ [numControllers - len(controllers)][len(controllers) - numControllers]int
+
+	// Watch the logging config map and dynamically update logging levels.
+	opt.ConfigMapWatcher.Watch(logging.ConfigMapName(), logging.UpdateLevelFromConfigMap(logger, atomicLevel, component))
```

Comment:
This needs `component` repeated at the end, although knative/pkg#436 will fix that.